src/fmtspec.jl

```diff
@@ -164,27 +165,38 @@ _srepr(x) = repr(x)
 _srepr(x::AbstractString) = x
 _srepr(x::AbstractChar) = string(x)
 _srepr(x::Enum) = string(x)
+@static if VERSION < v"1.2.0-DEV"
+    _srepr(x::Irrational{sym}) where {sym} = string(sym)
+end
 
 function printfmt(io::IO, fs::FormatSpec, x)
     cls = fs.cls
     ty = fs.typ
     if cls == 'i'
-        ix = Integer(x)
+        ix = x
+        try
+            ix = Integer(x)
+        catch
+        end
         ty == 'd' || ty == 'n' ? _pfmt_i(io, fs, ix, _Dec()) :
         ty == 'x' ? _pfmt_i(io, fs, ix, _Hex()) :
         ty == 'X' ? _pfmt_i(io, fs, ix, _HEX()) :
         ty == 'o' ? _pfmt_i(io, fs, ix, _Oct()) :
         _pfmt_i(io, fs, ix, _Bin())
     elseif cls == 'f'
-        fx = float(x)
+        fx = x
```

Comment:
Same as above, better for this to get an exception, instead of slowing this down by introducing type instability, and covering up what might be useful information for the programmer.

Reply:
I am not so sure whether we should care about type stability here. (I am definitely not an expert here!)
But packages like `PhysicalConstants` redefine `float()` to work on unitful quantities. So `float(x)` ends up to be something else than a subtype of AbstractFloat, e.g. `float(c_0)` is `2.99792458e8 m s^-1`.
So if we keep `fx = float(x)` , `fx` will be type unstable anyhow.

> 
> 
> > Well, I think I found a way. I am still not so familiar with Types...
> > still struggling to formulate something similar like `Complex{<:Integer}` if I remove the `Union{}`
> 
> Why do you need to separate out different types of complex numbers for defaults?
> For things like complex numbers, you'll need to be careful about how width / precision is handled,
> so that it doesn't just get treated like a string after the initial conversion to a string (which it looks like your code might be doing), which might end up having critical information truncated.

I think, you misinterpreted my code. I have added new less specific methods, e.g. `_pfmt_f(io, x::Number, ...)`, which are called if the currently existing method does not cover the type, i.e. if x is not a subtype of `AbstractFloat` in case of `_pfmt_f()`. The method calls the routine `_pfmt_Number_f()` which then calls `fmt_Number()`. (This method needs to be defined by the user, if he wants to add a new number type. For `Complex` types I have already defined it.) 

The last argument of that method is an anonymous formatting routine for the specific type ('i', 'e' or 'f') where the width of the `FormatSpec` is set to `-1`. This should be mainly a copy of the Number's `show()` method where the value is replaced by `f(value)`. The result type of the function needs to be a `String` which is then passed to `_pfmt_s()` with the original `FormatSpec`, which then takes into account the width of the resulting string.

Reply:
>The result type of the function needs to be a String which is then passed to _pfmt_s() with the original FormatSpec, which then takes into account the width of the resulting string.

That's the problem, because depending on the length, it may truncate important information.

Reply:
I can't see this: `f"\{1.3f}(1+2im)"` results in `"1.000 + 2.000im"`
`_pfmt_s()` only sets the minimal width.